File: algo_expert/BST/validate_BST.js
// when we explore tree.right, that node's upperbound is the current node
// when we explore tree.left, that node's lowerbound is the current node

// Time O(n), space O(D), D is the longest depth
function validateBst(tree) {
    return helper(tree, -Infinity, Infinity);
}

function helper(tree, minValue, maxValue) {
    if (!tree) return true;
    if (tree.value < minValue || tree.value >= maxValue) return false;

    let leftSide = helper(tree.left, minValue, tree.value);
    let rightSide = helper(tree.right, tree.value, maxValue);

    return leftSide && rightSide;
}

File: leet_code/python/binary_trees/validate_bst.py
# O(n) time | O(d) space
def validateBst(tree):
    return helper(tree, float("-inf"), float("inf"))


def helper(node, min_value, max_value):
	if not node:
		return True

	if node.value < min_value or node.value >= max_value:
		return False

	left = helper(node.left, min_value, node.value)
	right = helper(node.right, node.value, max_value)

	return left and right
